Add debugging doc (#330)
lsds/sgx-lkl: docs/Debugging.md

Debugging
=========

SGX-LKL provides a wrapper around gdb called `sgx-lkl-gdb`.
This wrapper is compatible with IDEs and can be used instead of `gdb`.
`sgx-lkl-gdb` automatically loads the SGX-LKL gdb plugin which
ensures that debug symbols (if available) are loaded correctly. 
When running in hardware mode, `sgx-lkl-gdb` uses the corresponding SGX debug
instructions to read from and write to enclave memory.

## Prerequisites

### Open Enclave

`sgx-lkl-gdb` relies on the Open Enclave gdb plugin.
[Install Open Enclave](https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/install_oe_sdk-Ubuntu_18.04.md) if necessary before continuing.

The instructions below assume that Open Enclave was installed in the default directory `/opt/openenclave`.

> If you used a different directory, run `. <OE_PREFIX>/share/openenclave/openenclaverc` first.

Reviewer: The statement implies we don't need to source `/opt/openenclave/share/openenclave/openenclaverc`.

Author: Yeah, that's the case. `sgx-lkl-gdb` checks the default directory. Should I mention that?

### Build mode

Debugging enclave code (SGX-LKL as well as the application code) is only possible in debug and non-release builds of SGX-LKL.

### Encrypted disks

> TODO add note on how to handle situations when disk keys are retrieved from a remote service that would not release them to debug/non-release builds of SGX-LKL (either re-create disks unencrypted, or, if available locally, hard-code key in app config for local debugging)

Reviewer: We should be able to debug the key release process. Ideally, the key vault should have a "debugger friendly" release policy. During dev/debug cycles, the disk encryption keys should be created with such policies.

Author: Agree, good point. I changed the TODO.

### Confidential Containers

The instructions below assume that disk images and configuration are available outside of a Confidential Container. Debugging of Confidential Containers (via Docker) is not supported currently.

## Option A: Debugging with an SGX-LKL installation

The instructions below assume that SGX-LKL was installed in `/opt/sgx-lkl`.

To debug an application, invoke `sgx-lkl-gdb` as follows:
```

> /opt/sgx-lkl/bin/sgx-lkl-gdb --args /opt/sgx-lkl/bin/sgx-lkl-run-oe --hw-debug root_disk.img /app

Reviewer: This is not accurate enough. It should be `<path-to-app>`, plus optionally parameters to `app`

Reviewer: The general idea is to prefix `sgx-lkl-gdb --args` to whatever command we used to launch the application in sgx-lkl.

Author: I changed the text and made it clear that the command is just an example.

```
See `sgx-lkl-run-oe --help` for how to run using host and app config files.

## Option B: Debugging with an SGX-LKL source build

This assumes you want to debug SGX-LKL directly from the source tree without installing it.

To debug an application, invoke `sgx-lkl-gdb` as follows:
```
tools/gdb/sgx-lkl-gdb --args build/sgx-lkl-run-oe --hw-debug root_disk.img /app
```
See `sgx-lkl-run-oe --help` for how to run using host and app config files.

## Debugging advice

### Visual Studio Code

Visual Studio Code allows to use a custom `gdb` command for debugging.
To use `sgx-lkl-gdb`, add `"miDebuggerPath": "/opt/sgx-lkl/bin/sgx-lkl-gdb"` (change path if necessary) in your `.vscode/launch.json` file.
Tip: Use the "Visual Studio Code Remote - SSH" extension to debug on a remote VM with SGX support.

### Run mode

> Apart from obvious differences (security, performance, ...) there are subtle differences between SGX-LKL software mode (`--sw-debug`) and hardware mode (`--hw-debug`).

Reviewer: It's probably simpler to say in `sw-debug` mode we simulate the SGX hardware.

Author: Changed, also added a concrete example of what isn't simulated currently.

If possible, debugging should be done in hardware mode to be as close to a production environment as possible.

### Ignore SIGILL

In some cases `sgx-lkl-gdb` still incorrectly stops at illegal instructions that are emulated inside the enclave (see e.g. https://github.com/lsds/sgx-lkl/issues/167). To avoid that, run `handle SIGILL pass nostop noprint` as first command in the active `gdb` session.

### Optimized code

In general, debugging optimized code is not recommended.
For an optimal debugging experience, SGX-LKL should be built in debug mode.
If the application code should be debugged as well and it is native code,

> then the application should be built without optimizations and with debug symbols.

Reviewer: We should mention somewhere at the top that debugging non-native applications is currently not possible. Debugging non-native language runtimes is also challenging because they typically are not built with symbols, e.g., https://github.com/lsds/sgx-lkl/issues/245

Author: I added a subsection at the top about non-native applications. I don't think we have to go into language runtimes as that's a very advanced topic and people would probably know what they are doing in that case.

Debugging
=========

SGX-LKL provides a wrapper around gdb called `sgx-lkl-gdb`.
This wrapper is compatible with IDEs and can be used instead of `gdb`.
`sgx-lkl-gdb` automatically loads the SGX-LKL gdb plugin which
ensures that debug symbols (if available) are loaded correctly. 
When running in hardware mode, `sgx-lkl-gdb` uses the corresponding SGX debug
instructions to read from and write to enclave memory.

## Prerequisites

### Open Enclave

`sgx-lkl-gdb` relies on the Open Enclave gdb plugin.
[Install Open Enclave](https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/install_oe_sdk-Ubuntu_18.04.md) if necessary before continuing.

The instructions below assume that Open Enclave was installed in the default directory `/opt/openenclave`.
If you used a different directory, run `. <OE_PREFIX>/share/openenclave/openenclaverc` first.

### Build mode

> Debugging enclave code (SGX-LKL as well as the application code) is only possible in debug and non-release builds of SGX-LKL.

Reviewer 2: ```suggestion
Note that debugging enclave code (SGX-LKL as well as application code) is not possible in RELEASE builds of SGX-LKL.
```

### Encrypted disks

TODO add note on how to handle situations when disk keys are retrieved from a remote service  (either re-create disks unencrypted, re-create disks encrypted but with debug-enabled key release policy (closest to production), or hard-code key in app config)

> 

Reviewer 2: I would remove this, as the details will depend on how key release is handled.

### Application type

> Debugging of non-native applications like Python, Java or .NET applications is currently not possible.

Reviewer 2: I don't understand what that means. What are "non-native applications"?

Author: Anything that runs in a language runtime, like Python scripts.

Reviewer 2: Ok, in that case, we should remove the statement that debugging of these applications is not possible. It actually depends how a debugger for a specific runtime works, and there is probably no point in going into that much detail here. (For example, debugging functionality that is part of an interpreter (or uses gdb underneath) may actually.) We should just claim `gdb` support and leave it at that.

### Docker deployment container

> The instructions below assume that disk images and configuration are available outside of a Docker deployment container. Debugging of such deployment containers is not supported currently as they do not contain the necessary debugging tools.

Reviewer 2: ```suggestion
The instructions below assume that disk images and the host/app configurations are available outside of a Docker deployment container. Debugging of deployment containers directly is not supported yet as they do not contain the necessary debugging tools.
```

See `sgx-lkl-run-oe --help` for how to run using host and app config files.

## Option B: Debugging with an SGX-LKL source build

This assumes you want to debug SGX-LKL directly from the source tree without installing it.

To debug an application, prefix the regular `sgx-lkl-run-oe` command with `sgx-lkl-gdb --args`, for example:
```
tools/gdb/sgx-lkl-gdb --args build/sgx-lkl-run-oe --hw-debug root_disk.img /app
```
See `sgx-lkl-run-oe --help` for how to run using host and app config files.

## Debugging advice

### Visual Studio Code

Visual Studio Code allows to use a custom `gdb` command for debugging.
To use `sgx-lkl-gdb`, add `"miDebuggerPath": "/opt/sgx-lkl/bin/sgx-lkl-gdb"` (change path if necessary) in your `.vscode/launch.json` file.
Tip: Use the "Visual Studio Code Remote - SSH" extension to debug on a remote VM with SGX support.

### Run mode

When using `--sw-debug` then SGX hardware is simulated. However, not all aspects are simulated, for example, certain CPU instructions are invalid within SGX, but would execute fine in simulation mode.
If possible, debugging should be done in hardware mode to be as close to a production environment as possible.

### Ignore SIGILL

> In some cases `sgx-lkl-gdb` still incorrectly stops at illegal instructions that are emulated inside the enclave (see e.g. https://github.com/lsds/sgx-lkl/issues/167). To avoid that, run `handle SIGILL pass nostop noprint` as first command in the active `gdb` session.

Reviewer 2: ```suggestion
In some cases `sgx-lkl-gdb` stops at illegal instructions that are emulated inside the enclave (see e.g. https://github.com/lsds/sgx-lkl/issues/167). To avoid that, run `handle SIGILL pass nostop noprint` as the first command in the active `gdb` session.
```
